common/quotas/clocked_rate_limiter_test.go

```diff
@@ -151,3 +151,81 @@ func TestClockedRateLimiter_Wait_DeadlineWouldExceed(t *testing.T) {
 	t.Cleanup(cancel)
 	assert.ErrorIs(t, rl.Wait(ctx), quotas.ErrRateLimiterReservationWouldExceedContextDeadline)
 }
+
+// test that reservations for 1 token ARE unblocked by RecycleToken
+func TestClockedRateLimiter_Wait_Recycle(t *testing.T) {
+	t.Parallel()
+	ts := clock.NewEventTimeSource()
+	rl := quotas.NewClockedRateLimiter(rate.NewLimiter(1, 1), ts)
+	ctx := context.Background()
+
+	// take first token
+	assert.NoError(t, rl.Wait(ctx))
+
+	// wait for next token and report when success
+	waiting := make(chan struct{})
+	done := make(chan struct{})
+	go func() {
```

Comment:
Do we need this outer go func()?

Reply:
yes, because we need to start waiting before we call recycle (otherwise recycle will be a no-op)

Reply:
time.Sleep(time.Millisecond) will ensure that it will wait before we call recycle right. I think that it will still work if you remove the outer goroutine and `waiting` channel.

Reply:
I was suggesting this:
```
// wait for next token and report when success
	done := make(chan struct{})
	go func() {
		assert.NoError(t, rl.Wait(ctx))
		done <- struct{}{}
	}()

	// The waiting channel isn't enough by itself to confirm waiting, because
	// the inner go func can be called without the rl.Wait() having finished.
	// Use time.Sleep to allow rl.Wait() to complete before signalling `waiting`.
	// 1-millisecond sleep succeeded >50,000x, whereas 1-nanosecond sleep deadlocks.
	time.Sleep(time.Millisecond)

	// once a waiter exists, recycle the token instead of advancing time
	rl.RecycleToken()

	// wait until done so we know assert.NoError was called
	<-done
```

Reply:
I agree the outer goroutine doesn't add anything here. the main goroutine just waits on waiting anyway

Reply:
I see, thank you! I was still thinking about the function without the sleep (but obviously the channel alone was not enough, so the sleep is doing the required work here)